fvm/evm/types/balance.go

```diff
@@ -18,50 +26,82 @@ var (
 // But on the FLOW Vaults, we use Cadence.UFix64 to store values in Flow.
 // this could result in accidental conversion mistakes, the balance object here would
 // do the conversions and does appropriate checks.
+type Balance struct {
+	Value *big.Int
+}
+
+// NewBalance constructs a new balance from flow value (how its stored in Cadence Flow)
+func NewBalance(inp cadence.UFix64) *Balance {
+	return &Balance{
+		Value: new(big.Int).Mul(
+			new(big.Int).SetUint64(uint64(inp)),
+			UFixToAttoConversionMultiplier),
+	}
+}
+
+// NewBalanceFromAttoFlow constructs a new balance from atto flow value
+func NewBalanceFromAttoFlow(inp *big.Int) *Balance {
+	return &Balance{
+		Value: inp,
+	}
+}
+
+// ToAttoFlow returns the balance in AttoFlow
+func (b *Balance) ToAttoFlow() *big.Int {
+	return b.Value
+}
+
+// Copy creates a copy of this balance
+func (b *Balance) Copy() *Balance {
+	return NewBalanceFromAttoFlow(b.ToAttoFlow())
+}
+
+// ToUFix64 casts the balance into a UFix64,
 //
-// For example the smallest unit of Flow token that a FlowVault could store is 1e10^-8,
+// Warning! The smallest unit of Flow token that a FlowVault (Cadence) could store is 1e10^-8,
 // so transfering smaller values (or values with smalls fractions) could result in loss in
-// conversion. The balance object checks it to prevent invalid balance.
-// This means that values smaller than 1e10^-8 flow could not be bridged between FVM and Flow EVM.
-type Balance cadence.UFix64
+// conversion. The rounded flag should be used to prevent loss of assets.
+func (b *Balance) ToUFix64() (cadence.UFix64, error) {
+	var err error
+	converted := new(big.Int).Div(b.Value, UFixToAttoConversionMultiplier)
+	if !converted.IsUint64() {
+		// this should never happen
+		err = fmt.Errorf("balance can't be casted to a uint64")
+	}
+	return cadence.UFix64(converted.Uint64()), err
+}
 
-// ToAttoFlow converts the balance into AttoFlow
-func (b Balance) ToAttoFlow() *big.Int {
-	return new(big.Int).Mul(new(big.Int).SetUint64(uint64(b)), SmallestAcceptableBalanceValueInAttoFlow)
+// HasUFix64RoundingError returns true if casting to UFix64 has rounding error
+func (b *Balance) HasUFix64RoundingError() bool {
+	return new(big.Int).Mod(b.Value, UFixToAttoConversionMultiplier).Cmp(big.NewInt(0)) != 0
 }
 
 // Sub subtract the other balance from this balance
-func (b Balance) Sub(other Balance) Balance {
-	// no need to check for underflow, as go does it
-	return Balance(uint64(b) - uint64(other))
+func (b *Balance) Sub(other *Balance) error {
+	otherInAtto := other.ToAttoFlow()
+	// check underflow b < other
+	if b.Value.Cmp(otherInAtto) == -1 {
+		return ErrWithdrawBalanceRounding
+	}
+	b.Value = new(big.Int).Sub(b.Value, otherInAtto)
```

Comment:
this result could be a value that produces a rounding error when converted to Flow no?